### ISSUE TITLE
Memory Leaks in Jasmine

### ISSUE THREAD

**Issue by fracmak** [jasmine/jasmine#1154]
So I'm working on a large angular project with a large number of tests, 2539 as of writing this. In my tests, I'm using angular mock to inject angular variables into my tests in a beforeEach() and then saving them in a scoped variable for my tests to use. The problem I'm having is that jasmine maintains various references to the it(), describe(), beforeEach(), afterEach() functions inside it's suite object (https://github.com/jasmine/jasmine/blob/master/lib/jasmine-core/jasmine.js#L2186-L2192) which prevent any garbage collection from occurring of these scoped variables, so currently the memory footprint just spikes up to 2+ gigs of memory when we finally reach some internal limit of the browser's memory footprint and we get a browser disconnect. 

My current solution is to have an afterEach() in each test file where it clears out these scoped variables, but it would be great if jasmine released it's references to the tests after they were run.

**Comment by amavisca** [comment]
This might not "solve" your issue as you've a large existing codebase, but I recommend not capturing variables in the closure to avoid these memory issues. Jasmine hands you a fresh `this` object inside of each spec run that gets wiped out between every `it`.

Stumbled upon this nice writeup that I hadn't seen before by some Jasmine users: https://gist.github.com/traviskaufman/11131303

There's some related discussion on this memory issue here: https://github.com/jasmine/jasmine/issues/941

**Comment by mikeperri** [comment]
Is there any reason Jasmine can't just delete the references to the functions after they're executed? 

I made a function like this:
```
    var clearSuiteFns = function(suite) {
        function clearArray(arr, message) {
          for (var i = 0; i < arr.length; i++) {
            delete arr[i].fn;
          }
        }

        clearArray(suite.afterAllFns);
        clearArray(suite.afterFns);
        clearArray(suite.beforeAllFns);
        clearArray(suite.beforeFns);
    };
```


and then called it after each suite finishes:
```
getJasmineRequireObj().Env = function(j$) {
  function Env(options) {
...
    this.execute = function(runnablesToRun) {
...
      var processor = new j$.TreeProcessor({
...
        nodeComplete: function(suite, result) {
          clearSuiteFns(suite);

          if (!suite.disabled) {
            clearResourcesForRunnable(suite.id);
          }
          currentlyExecutingSuites.pop();
          reporter.suiteDone(result);
        },
...
      });
...
  }
};
```

and then I just added a line to delete queueableFn's fn when a spec completes:
```
getJasmineRequireObj().Spec = function(j$) {
...
  Spec.prototype.execute = function(onComplete, enabled) {
...
    function complete(enabledAgain) {
      self.result.status = self.status(enabledAgain);
      self.resultCallback(self.result);
      delete self.queueableFn.fn;

      if (onComplete) {
        onComplete();
      }
    }
...
  }
}
```
Here's my commit:
https://github.com/mikeperri/jasmine/commit/bb9adf60e777665fd57395a24a5fc63cd727b19a

All the tests pass - is there any reason why this would cause an issue? Is there ever some scenario where a suite or spec might need to be run multiple times?



I tried to add some unit & integration tests but I run into some issues, namely, neither of those complete callbacks seem to be getting called by jasmineUnderTest.

**Comment by fracmak** [comment]
@mikeperri that is the type of solution I've been looking for. I tried doing something similar but it didn't make a big difference in the memory footprint. I'll try your code and let you know how well it works

**Comment by mikeperri** [comment]
@fracmak how did it go?

**Comment by fracmak** [comment]
sadly it didn't make that big a dent

**Comment by tpflueger** [comment]
@fracmak I've also had issues. The project I was on had close to 6k tests. Angular came out with a way to limit the amount of times the app is initialized, which greatly decreased some of the memory issues we were having.

https://docs.angularjs.org/api/ngMock/function/angular.mock.module.sharedInjector

**Comment by slackersoft** [comment]
We've looked at deleting functions once Jasmine is done with them, in order to clean out anything that is closed over (See https://www.pivotaltracker.com/story/show/56030214). Unfortunately, there are users who expect to be able to `execute` a Jasmine environment multiple time without reloading the specs, and this isn't something we're quite ready to break compatibility for.

For now, you should probably use the `this` context that is shared between the `beforeEach`s, `it`, and `afterEach`s and inherited from `beforeAll`s and `afterAll`s. It will be available for garbage collection once it is no longer in use by your specs.

Closing. Thanks for using Jasmine!

**Comment by thorn0** [comment]
> Unfortunately, there are users who expect [...], and this isn't something we're quite ready to break compatibility for.

@slackersoft Sounds like this behavior simply should be optional. What are your thoughts on introducing an option for this?

**Comment by slackersoft** [comment]
Jasmine 3.0 should be `null`ing out references to functions that Jasmine has completed execution of. This should allow variables that were closed over by specs, befores, and afters to be cleaned up by the javascript runtime. If you're still seeing further leaks, we should open a new issue with some more details of what kinds of things are leaking, so we can figure out strategies to fix them.

Hope this helps. Thanks for using Jasmine!
